### ISSUE TITLE
How to get untested files to show in the report as 0% coverage in a cross compilation setup?

### ISSUE THREAD

**Issue by rds-tmeyer** [gcovr/gcovr#536]
Hi,

I'm using gcovr to generate unit test report for a cross compiled firmware project. Since I am still building the unit test cases, I would like to show file that have no test case as not covered (yet).

Here is my project structure : 
```
/MyProj
├── src
│   ├── module1.c
│   ├── module1.h
│   ├── module2.c
│   ├── module2.h
│   ├── module3.c
│   └── module3.h
└── tests
    ├── Makefile
    ├── Module1
    │   ├── Makefile
    │   └── test_module1.c
    ├── Module2
    │   ├── Makefile
    │   └── test_module2.c
    └── Module3
        ├── Makefile
        └── test_module3.c
```

When I run the Makefile in each module test subdir, I create a report with only the coverage for the module in question, which is fine. When I call the global Makefile, I'm using the --add-trace feature to combine the individual report (in json format) into a global one. I would like this global report to also show the uncovered files from the src/ directory.

Each test is cross compiled in a single executable and generates the gcda and gcno files for the given module, but it does not generate gcno files for other module, because the other module are not compiled for a given module test.

Here are the command I'm using : 
1. to generate module related coverage (this is called next to test_moduleX.c : /MyProj/tests/ModuleX)
`	gcovr -r "/MyProj/src" --filter ".+\.c$" --json-pretty -o "results/coverage.json" .
` 
2. to generate project related coverage (run in /MyProj/tests) : I would like this one to include uncovered file
`	gcovr -r "/MyProj/src" --add-tracefile "/MyProj/tests/*/coverage.json" --html-details global_coverage.html
` 

How can this be achieved ?

**Comment by latk** [comment]
Gcovr only knows about files that were included in a compilation unit. But you don't have to link the code into your executable. It should be sufficient when you just compile the other modules to an object file so that a .gcno file exists, and make sure that these files are in a path that gcovr searches.

So I think you could do the following:

* compile all modules into object files with coverage, so that you get `moduleX.o` and `moduleX.gcno` somewhere
* no need to run an executable to create gcda files
* run gcovr so that it processes those gcno files, creating a coverage report in which everything is uncovered
* merge your reports as desired, including this baseline in which nothing was covered

This will not reduce your coverage percentage. A coverable object (line/branch) counts as fully covered if it was executed at least once.

Also, thank you for asking a very clear question with all the necessary paths!

**Comment by rds-tmeyer** [comment]
It's a good idea but compiling all the untested module is not possible because they require stubbing and/or mocking. This is done only when the test case is being written. I would like to get an overview of the coverage before starting this stubbing effort. Would it be possible to create some dummy gcno files from the source ?

**Comment by latk** [comment]
Only the compiler can create the gcno files. If you just want empty files to appear in your coverage report you could also create a JSON file with the suitable format for feeding into gcovr:

```json
{
  "gcovr/format_version": ...,
  "files": [
    { "file": "module1.c", "lines": [] },
    ...
    { "file": "moduleN.c", "lines": [] }
  ]
}
```

However, this will not produce proper line-coverage and branch-coverage metrics. To get those, it is not possible to circumvent the compiler.

But again: you only have to compile the modules, not get them into a runnable form. You don't have to link any libraries. You don't have to supply functional mocks.

You might be able to use gcno files created for a different architecture. Gcovr itself is agnostic to cross-compilation issues, it just requires that your gcov program matches your compiler toolchain (see the `--gcov-executable` command line option). Moving coverage data between multiple systems is supported by GCC/gcov, but can be very tricky as the files contain file paths from the system they were compiled on, see [Cross-Profiling in the GCC docs](https://gcc.gnu.org/onlinedocs/gcc/Cross-profiling.html#Cross-profiling). If at all possible, avoid moving the gcno files. But once you've turned that into a gcovr JSON file, you can freely merge the data with other reports (as long as all JSON files were generated with the same root directory).

**Comment by latk** [comment]
I'm closing this issue because gcovr is behaving as expected. I hope one of the suggested workarounds helped.

**Comment by rds-tmeyer** [comment]
I'm still trying to figure out the best way but indeed it's not an issue with gcovr.

**Comment by prayassamriya** [comment]
Hi @rds-tmeyer, We are looking for something similar. we started with a unit test activity for our big project and its going to be a monster work to get all units added into unit test project. To show true coverage for whole project, we are looking for a solution which can provide us coverage of all files into unit test project without the need to really added those into unit test project ( as it will require efforts to create stubs and fakes to get rid of external dependencies ). The project is based on zephyr rtos using STM32 as targets ( and gcc compiler ) where as unit test project uses google test framework to be run on windows/linux host env.
while we know that coverage for all those files ( not part of the unit test project ) will be 0 percent however how to automate this process and get something in the form of generatable report.
please let me know if you ended up with a solution which could meet our requriements

**Comment by Spacetown** [comment]
You can generate the file mentioned in https://github.com/gcovr/gcovr/issues/536#issuecomment-993009205 by using a shell script with a file list:

```shell
files_lst=/tmp/files.lst
(
    echo "file1.c"
    echo "dir/file2.cpp"
) > $files_lst

(
   echo '''
{
  "gcovr/format_version": ...,
  "files": [
'''
   sed -e 's/\(.*\)/{ "file": "\1", "lines": [] }/; 2, $ { s/\(.*\)/, \1/; }' < $files_lst
   echo '''
  ]
}
'''
)
```

**Comment by prayassamriya** [comment]
thanks, we will try this method.
